### PR TITLE
feat(es/base): Add `TransformConext` to reduce breaking changes

### LINKED ISSUE #8186: `swc_ecma_transforms_base::TransformContext`
All transform functions for compatibility and bugfixes should accept it.

This can be a bit cumbersome, but this will reduce breaking changes due to more assumption required.

```rust

pub struct TransformContext<C> {
    pub comments: C,
    pub assumptions: Assumptions,
    pub unresolved_mark: Mark.
}


```

I'm not sure about the extra fields

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

This will force users to create needlessly many input, but this will reduce breaking change.
I want to create an `swc-rs-ecosystem-ci` and this is groundwork for it.

**Related issue:**

 - Closes #8186

### CODE REVIEW

File: crates/swc_ecma_compat_es2015/src/block_scoping/mod.rs [hunk 1]
```diff
@@ -37,7 +38,14 @@ mod vars;
 ///    });
 /// }
 /// ```
-pub fn block_scoping(unresolved_mark: Mark) -> impl VisitMut + Fold {
+pub fn block_scoping<C>(
```
Comment (kdy1) [inline crates/swc_ecma_compat_es2015/src/block_scoping/mod.rs:41]: This became much verbose

File: crates/swc_ecma_compat_es2015/src/classes/mod.rs [hunk 2]
```diff
@@ -31,14 +30,25 @@ use self::{
 mod constructor;
 mod prop_name;
 
-pub fn classes<C>(comments: Option<C>, config: Config) -> impl Fold + VisitMut
+pub fn classes<C>(
```
Comment (kdy1) [inline crates/swc_ecma_compat_es2015/src/classes/mod.rs:33]: This is actually cleaner, and it also allows future optimization

File: crates/swc_ecma_compat_es2015/src/classes/mod.rs [hunk 3]
```diff
@@ -31,14 +30,25 @@ use self::{
...
+) -> impl Fold + VisitMut
 where
-    C: Comments,
+    C: Comments + Clone,
 {
     as_folder(Classes {
         in_strict: false,
         comments,
-        config,
+        config: Config {
```
Comment (kdy1) [inline crates/swc_ecma_compat_es2015/src/classes/mod.rs:46]: The constructor drops `Assumptions` early to avoid making difference of memory access

**Review by kdy1 — COMMENTED** [review]
I want to hear opinions from others.

The problem I'm trying to solve is `swc-rs-ecosystem-ci`, which is about slowing the release process of `swc_core` down and preventing regression of it.

But swc exposes an enormous amount of API, especially for transforms. This means preventing a breaking change is hard. To reduce the direct API surface, I'm experimenting with a type `TransformContext`, which should be passed to almost all transforms. Because each transform reads relevant **assumptions** from the context, you don't need to pass fields of assumptions like at the moment.

This will prevent breaking changes due to added optimizations, where a new assumption is required.

Of course, this is not enough to prevent all breaking changes, and I'll make other PRs to reduce the API surface, and breaking chanages.

# Alternatives

## `#[non_exhaustive]` on every config stfuct.


---

Pros:
 - It's not a breaking change if we implement a new optimization using a new assumption.
 - Users do not need to create read API docs to figure out the types.

Cons:
 - APIs become compilcated.
 - A user has to create `TransformContext` even if it's not going to be used at all.

**Comment by kdy1** [comment]
cc @kwonoj @magic-akari @Austaras

File: crates/swc_ecma_compat_es2015/src/arrow.rs [hunk 4]
```diff
@@ -57,7 +58,14 @@ use swc_trace_macro::swc_trace;
 /// };
 /// console.log(bob.printFriends());
 /// ```
-pub fn arrow(unresolved_mark: Mark) -> impl Fold + VisitMut + InjectVars {
+pub fn arrow<C>(
```
Comment (magic-akari) [inline crates/swc_ecma_compat_es2015/src/arrow.rs:61]: Does this mean that every visitor have to add a generic parameter of type C?

**Review by magic-akari — DISMISSED** [review]
We should do it

File: crates/swc_ecma_compat_es2015/src/arrow.rs (hunk 4 above)
Comment (kdy1) [reply crates/swc_ecma_compat_es2015/src/arrow.rs:61]: Yes, at least all visitors accept `TransformContext`

**Comment by Austaras** [comment]
Do people really relies on separate compat passes? Maybe only change esxxx and preset-env would be enough?

**Comment by kdy1** [comment]
Some do, but I think it makes sense

**Comment by kdy1** [comment]
But then the problem is that it's a _technically_ breaking change.

**Pull request closed without merging**